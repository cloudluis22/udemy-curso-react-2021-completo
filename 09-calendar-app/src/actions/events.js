import { __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED } from "react-dom";
import Swal from "sweetalert2";
import { convertEvents } from "../helpers/convertEvents";
import { fetchToken } from "../helpers/fetch";
import { types } from "../types/types";

export const eventStartAddNew = (event) => {
    return async ( dispatch, getState ) => {

        const { uid, name } = getState().auth;

        try {
            const dbEvent = {
                ...event,
                start: event.startDate,
                end: event.endDate
            }

            const response = await fetchToken('events', dbEvent, 'POST');
            const body = await response.json();

            if (body.ok) {
                event.id = body.event.id
                event.user = {
                    _id: uid,
                    name: name
                }
                console.log(event);
                dispatch( eventAddNew(event) );
            }

        } catch (error) {
            console.log(error)
        }

    };
};

export const eventStartLoading = () => {
    return async (dispatch) => {

        try {
            const response = await fetchToken('events');
            const body = await response.json();

            const events = convertEvents(body.events);
           dispatch( eventsLoaded( events ) );

        } catch (error) {

        }

    };
};

export const eventStartUpdate = (event) => {
    return async (dispatch) => {

        try {

            const response = await fetchToken(`events/${event.id}`, event, 'DELETE');
            const body = await response.json();

            if (body.ok) {
                dispatch( eventUpdated(event) );
            }
            else {
                Swal.fire('Error', body.msg, 'error');
            }

        } catch (error) {
            console.log(error);
        };

    };
};

export const eventStartDelete = () => {


    return async (dispatch, getState) => {
        try {

            const { id } = getState().calendar.activeEvent;

            const response = await fetchToken(`events/${id}`, {}, 'delete');
            const body = await response.json();

            if (body.ok) {
                dispatch( eventDeleted() );
            }
            else {
                Swal.fire('Error', body.msg, 'error');
            }

        } catch (error) {
            console.log(error);
        };

    };
}

export const eventLogoutCleaning = () => ({ type: types.eventLogoutCleaning });

const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActive = () => ({
    type: types.eventClearActive
});

const eventUpdated = ( event ) => ({
    type: types.eventUpdated,
    payload: event
});

const eventDeleted = () => ({
    type: types.eventDeleted
});

const eventsLoaded = (events) => ({ type: types.eventsLoaded, payload: events });