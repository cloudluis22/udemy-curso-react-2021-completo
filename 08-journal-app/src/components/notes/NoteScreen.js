import React, { useEffect, useRef } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';
import { activeNote, startDeleting } from '../../actions/notes'

export const NoteScreen = () => {

    const dispatch = useDispatch();

    const { active: actNote } = useSelector(state => state.notes);
    const [ formValues, handleInputChange, reset ] = useForm(actNote);

    const { body, title } = formValues;

    const activeId = useRef( actNote.id );

    useEffect( () => {

        if (actNote.id !== activeId.current) {
            reset(actNote);
            activeId.current = actNote.id;
        }

    }, [actNote, reset]);

    useEffect(() => {

        dispatch( activeNote( formValues.id, { ...formValues } ) ) ;

    }, [formValues, dispatch])

    const handleDelete = () => {

        dispatch( startDeleting( activeId.current ) );
    }

    return (
        <div className="notes__main-content">
            <NotesAppBar />

            <div className="notes__content">
                <input
                    type="text"
                    placeholder="Put a title here"
                    className="notes__title-input"
                    autoComplete="off"
                    name="title"
                    value={ title }
                    onChange={ handleInputChange }
                />
                <textarea
                    placeholder="What happened today?"
                    className="notes__textarea"
                    name="body"
                    value={body}
                    onChange={ handleInputChange }
                >
                    </textarea>

                {
                    (actNote.url) &&
                    <div className="notes__image">
                        <img
                            src={ actNote.url }
                            alt="imagen"
                        />
                    </div>
                }

            </div>

                <button className="btn btn-danger" onClick={ handleDelete } > Delete </button>

        </div>
    )
}