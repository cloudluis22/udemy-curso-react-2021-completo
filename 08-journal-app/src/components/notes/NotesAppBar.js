import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';
import { activeNote, startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {

    const dispatch = useDispatch();
    const { active } = useSelector( state => state.notes );
    const date = moment(active.date);

    const handleSave = () => {
        dispatch( startSaveNote(active) );
    }

    const handlePictureUpload = () => {
        document.querySelector('#fileSelector').click();
    }

    const handleFileChange = (evt) => {

        const file = evt.target.files[0];

        if (file) {
            dispatch( startUploading(file) );
        }

    }

    return (
        <div className="notes__appbar">
            <span> { date.format('ll') } </span>

            <input id="fileSelector" type="file" style={ { display: 'none' } } onChange={ handleFileChange } />

            <div>
                <button className="btn" onClick={ handlePictureUpload } > Picture </button>
                <button className="btn" onClick={ handleSave }> Save </button>
            </div>
        </div>
    )
}
