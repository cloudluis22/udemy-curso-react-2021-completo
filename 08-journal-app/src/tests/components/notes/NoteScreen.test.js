import React from 'react';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';

import { NoteScreen } from '../../../components/notes/NoteScreen';
import { activeNote } from '../../../actions/notes';

jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        active: {
            id: '123456',
            title: 'hola',
            body: 'mundo',
            date: 221455365
        },
        notes: []
    }
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(

    <Provider store={store}>
        <NoteScreen />
    </Provider>
);

describe('Pruebas en el componente funcional de "<NoteScreen />."', () => {

    test('Debe de renderizarse correctamente y hacer match con el snapshot. ', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('La acciÃ³n de "activeNote" debe dispararse correctamente.', () => {

        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'orozco se la come'
            }
        });

        expect(activeNote).toHaveBeenCalled();

    });

});