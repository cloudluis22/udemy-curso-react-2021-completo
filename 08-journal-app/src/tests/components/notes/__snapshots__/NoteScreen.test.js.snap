// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente funcional de "<NoteScreen />." Debe de renderizarse correctamente y hacer match con el snapshot.  1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
       
      Jan 3, 1970
       
    </span>
    <input
      id="fileSelector"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
         Picture 
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
         Save 
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Put a title here"
      type="text"
      value="hola"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today?"
      value="mundo"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
     Delete 
  </button>
</div>
`;
